### ISSUE TITLE
[question] importing xquery module from .jar file

### ISSUE THREAD

**Issue by wsdookadr** [BaseXdb/basex#1363]
Hi,

I'm interested in packaging an application as a standalone .jar
The app makes use of BaseX and therefore it has some XQuery programs
it comes with. Currently I load those like this from Java and then execute them on
the BaseX server:

    Class.class.getClass().getResourceAsStream("/path/to/resource")

This is how my resources directory looks like:

    $ tree resources/
    resources/
    ├── add_checksum.xql
    ├── create_db.xql
    ├── create_doc.xql
    ├── create_schema.sql
    ├── fcall_check.xql
    ├── insert_checksum.sql
    ├── mast.xql
    ├── strings.xql
    ├── test.xql
    └── var_check.xql

I'd like to restructure these into XQuery modules, but I'm not sure how an `import module` statement would look like if I would import a file that's inside the `.jar` file.

I also skimmed through [this email thread](https://mailman.uni-konstanz.de/pipermail/basex-talk/2015-April/thread.html#8613) on the basex mailing list, and [the docs](http://docs.basex.org/wiki/Repository) but wasn't able to find the answer there.

Is it currently possible in BaseX to import a module from a classpath URL?

As a fallback, if this is not possible, I might copy all the resources from the .jar
somewhere on the disk, but I'd like to avoid this.

tl;dr what are some options for importing an xquery module in an  xquery program when both of them are contained in a Java .jar file

**Comment by ChristianGruen** [comment]
Hi Stefan, currently, there is no solution for that. I had another look at the thread you mentioned, and it seems that, back then, the discussion stopped at that point. So… Suggestions how this could be done are welcome ;)

**Comment by wsdookadr** [comment]
Hey @ChristianGruen , Thank you for your answer

Currently, I would be interested how this is done in existing projects that make use of BaseX. Perhaps pointers to some codebases on github that have multiple such `.xql` and `.xqm` files. Or, what your prefered approach in the context of the functionality present in the latest released version.

**Comment by ChristianGruen** [comment]
In our own projects, we usually proceed as follows:

* Are files are deployed in a WAR file and unzipped by the web server, or
* the full distribution of BaseX is deployed and unzipped on a remote server (incl. Jetty)

**Comment by wsdookadr** [comment]
Thanks! yeah, I might eventually rethink the way I'm doing this.

Would evaluating the `.xqm` module using [`xquery:eval`](http://docs.basex.org/wiki/XQuery_Module#xquery:eval) and then running the actual `.xql` work?
I mean as an alternative.

**Comment by ChristianGruen** [comment]
I found that, in my last reply, I focused on Web Applications. One more note that could be helpful: Usually, we place our library modules in the project’s `repo` directory (see http://docs.basex.org/wiki/Repository#Accessing_Modules); this way, there is no need to specify the location of a module in the import statement.

> Would evaluating the `.xqm` module using [`xquery:eval`](http://docs.basex.org/wiki/XQuery_Module#xquery:eval) and then running the actual `.xql` work?

I’m not sure if I can follow you… Could you give some more information?

**Comment by wsdookadr** [comment]
> I’m not sure if I can follow you… Could you give some more information?

Yes, this is what I had in mind:

```java
String module = Class.class.getClass().getResourceAsStream("/path/to/module.xqm")
String query = Class.class.getClass().getResourceAsStream("/path/to/module.xql")
session.execute(String.format("xquery:eval(%s)",module));
session.execute(query);
```

**Comment by ChristianGruen** [comment]
Sorry, I guess no… Both strings would be executed as separate queries. – Are you working with local or remote sessions?

**Comment by ChristianGruen** [comment]
Btw, for future requests: We use GitHub mostly for fixing bugs. The best platform for questions is our mailing list. Thanks!

**Comment by ChristianGruen** [comment]
I’m closing this for now, in favor of our mailing list… Hope it’s ok ;)

**Comment by wsdookadr** [comment]
That's fine, thanks!
I will reopen this on the mailing list later on with more details
